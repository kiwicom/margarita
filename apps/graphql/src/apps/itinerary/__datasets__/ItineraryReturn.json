{
  "ref_tasks": [],
  "search_params": {
    "to_type": "airport",
    "flyFrom_type": "airport",
    "seats": {
      "infants": 0,
      "passengers": 1,
      "adults": 1,
      "children": 0
    }
  },
  "currency_rate": 1,
  "refresh": [],
  "connections": [],
  "currency": "EUR",
  "del": null,
  "all_airlines": [],
  "time": 1,
  "all_stopover_airports": [],
  "data": [
    {
      "quality": 1254.798848,
      "flyTo": "LIM",
      "deep_link": "https://www.kiwi.com/deep?from=OSL&to=LIM&departure=05-05-2019&return=06-06-2019&flightsId=8903845177446804_0%7C19912156091999362_0%7C19912156091999362_1%7C19912156091999362_2%7C19912156091999362_3%7C4050602064459815_0&price=794&passengers=1&affilid=kiwicomprgafilltest&lang=en&currency=EUR&booking_token=token",
      "nightsInDest": 32,
      "airlines": [
        "LA",
        "DY"
      ],
      "pnr_count": 3,
      "baglimit": {
        "hand_width": 23,
        "hold_weight": 20,
        "hand_weight": 7,
        "hand_height": 35,
        "hand_length": 55
      },
      "has_airport_change": false,
      "distance": 11079.54,
      "type_flights": [
        "GDS-r",
        "lcc-DY"
      ],
      "bags_price": {
        "1": 160.19
      },
      "flyFrom": "OSL",
      "booking_token": "token",
      "cityFrom": "Oslo",
      "duration": {
        "total": 196560,
        "return": 111960,
        "departure": 84600
      },
      "id": "8903845177446804_0|19912156091999362_0|19912156091999362_1|19912156091999362_2|19912156091999362_3|4050602064459815_0",
      "conversion": {
        "EUR": 794
      },
      "countryTo": {
        "code": "PE",
        "name": "Peru"
      },
      "price": 794,
      "routes": [
        [
          "OSL",
          "LIM"
        ],
        [
          "LIM",
          "OSL"
        ]
      ],
      "cityTo": "Lima",
      "transfers": [],
      "route": [
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 1738,
          "operating_carrier": null,
          "cityTo": "Barcelona",
          "vehicle_type": "aircraft",
          "flyFrom": "OSL",
          "id": "8903845177446804_0",
          "equipment": null,
          "combination_id": "8903845177446804",
          "fare_family": "",
          "flyTo": "BCN",
          "airline": "DY",
          "fare_classes": "",
          "cityFrom": "Oslo",
          "last_seen": "2019-04-03T22:46:12.000Z",
          "guarantee": false,
          "fare_basis": "",
          "local_arrival": "2019-05-05T09:50:00.000Z",
          "utc_arrival": "2019-05-05T07:50:00.000Z",
          "local_departure": "2019-05-05T06:30:00.000Z",
          "utc_departure": "2019-05-05T04:30:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": true,
          "return": 0,
          "flight_no": 2431,
          "operating_carrier": "LA",
          "cityTo": "Lima",
          "vehicle_type": "aircraft",
          "flyFrom": "BCN",
          "id": "19912156091999362_0",
          "equipment": "763",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "LIM",
          "airline": "LA",
          "fare_classes": "A",
          "cityFrom": "Barcelona",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": true,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-05-05T23:00:00.000Z",
          "utc_arrival": "2019-05-06T04:00:00.000Z",
          "local_departure": "2019-05-05T16:50:00.000Z",
          "utc_departure": "2019-05-05T14:50:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 1321,
          "operating_carrier": "LA",
          "cityTo": "Asunci처n",
          "vehicle_type": "aircraft",
          "flyFrom": "LIM",
          "id": "19912156091999362_1",
          "equipment": "320",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "ASU",
          "airline": "LA",
          "fare_classes": "N",
          "cityFrom": "Lima",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-06T05:04:00.000Z",
          "utc_arrival": "2019-06-06T09:04:00.000Z",
          "local_departure": "2019-06-06T00:29:00.000Z",
          "utc_departure": "2019-06-06T05:29:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 1317,
          "operating_carrier": "LA",
          "cityTo": "S찾o Paulo",
          "vehicle_type": "aircraft",
          "flyFrom": "ASU",
          "id": "19912156091999362_2",
          "equipment": "320",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "GRU",
          "airline": "LA",
          "fare_classes": "B",
          "cityFrom": "Asunci처n",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-06T13:40:00.000Z",
          "utc_arrival": "2019-06-06T16:40:00.000Z",
          "local_departure": "2019-06-06T10:30:00.000Z",
          "utc_departure": "2019-06-06T14:30:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 8114,
          "operating_carrier": "LA",
          "cityTo": "Barcelona",
          "vehicle_type": "aircraft",
          "flyFrom": "GRU",
          "id": "19912156091999362_3",
          "equipment": "359",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "BCN",
          "airline": "LA",
          "fare_classes": "A",
          "cityFrom": "S찾o Paulo",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-07T08:45:00.000Z",
          "utc_arrival": "2019-06-07T06:45:00.000Z",
          "local_departure": "2019-06-06T17:20:00.000Z",
          "utc_departure": "2019-06-06T20:20:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": true,
          "return": 1,
          "flight_no": 1739,
          "operating_carrier": null,
          "cityTo": "Oslo",
          "vehicle_type": "aircraft",
          "flyFrom": "BCN",
          "id": "4050602064459815_0",
          "equipment": null,
          "combination_id": "4050602064459815",
          "fare_family": "",
          "flyTo": "OSL",
          "airline": "DY",
          "fare_classes": "",
          "cityFrom": "Barcelona",
          "last_seen": "2019-04-04T06:10:05.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-06-07T14:35:00.000Z",
          "utc_arrival": "2019-06-07T12:35:00.000Z",
          "local_departure": "2019-06-07T11:10:00.000Z",
          "utc_departure": "2019-06-07T09:10:00.000Z"
        }
      ],
      "facilitated_booking_available": true,
      "countryFrom": {
        "code": "NO",
        "name": "Norway"
      },
      "local_arrival": "2019-05-05T23:00:00.000Z",
      "utc_arrival": "2019-05-06T04:00:00.000Z",
      "local_departure": "2019-05-05T06:30:00.000Z",
      "utc_departure": "2019-05-05T04:30:00.000Z"
    },
    {
      "quality": 1299.798858,
      "flyTo": "LIM",
      "deep_link": "https://www.kiwi.com/deep?from=OSL&to=LIM&departure=05-05-2019&return=06-06-2019&flightsId=3991228399399864_0%7C19912156091999362_0%7C19912156091999362_1%7C19912156091999362_2%7C19912156091999362_3%7C4050602064459815_0&price=843&passengers=1&affilid=kiwicomprgafilltest&lang=en&currency=EUR&booking_token=token",
      "nightsInDest": 32,
      "airlines": [
        "SK",
        "DY",
        "LA"
      ],
      "pnr_count": 3,
      "baglimit": {
        "hand_width": 23,
        "hold_weight": 20,
        "hand_weight": 7,
        "hand_height": 35,
        "hand_length": 55
      },
      "has_airport_change": false,
      "distance": 11079.54,
      "type_flights": [
        "GDS-r",
        "lcc-DY",
        "lcc-SK"
      ],
      "bags_price": {
        "1": 223.53
      },
      "flyFrom": "OSL",
      "booking_token": "token",
      "cityFrom": "Oslo",
      "duration": {
        "total": 194760,
        "return": 111960,
        "departure": 82800
      },
      "id": "3991228399399864_0|19912156091999362_0|19912156091999362_1|19912156091999362_2|19912156091999362_3|4050602064459815_0",
      "conversion": {
        "EUR": 843
      },
      "countryTo": {
        "code": "PE",
        "name": "Peru"
      },
      "price": 843,
      "routes": [
        [
          "OSL",
          "LIM"
        ],
        [
          "LIM",
          "OSL"
        ]
      ],
      "cityTo": "Lima",
      "transfers": [],
      "route": [
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": false,
          "return": 0,
          "flight_no": 4685,
          "operating_carrier": "SK",
          "cityTo": "Barcelona",
          "vehicle_type": "aircraft",
          "flyFrom": "OSL",
          "id": "3991228399399864_0",
          "equipment": "73H",
          "combination_id": "3991228399399864",
          "fare_family": "",
          "flyTo": "BCN",
          "airline": "SK",
          "fare_classes": "L",
          "cityFrom": "Oslo",
          "last_seen": "2019-04-03T23:27:37.000Z",
          "guarantee": false,
          "fare_basis": "LNOLIGHT",
          "local_arrival": "2019-05-05T10:25:00.000Z",
          "utc_arrival": "2019-05-05T08:25:00.000Z",
          "local_departure": "2019-05-05T07:00:00.000Z",
          "utc_departure": "2019-05-05T05:00:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": true,
          "return": 0,
          "flight_no": 2431,
          "operating_carrier": "LA",
          "cityTo": "Lima",
          "vehicle_type": "aircraft",
          "flyFrom": "BCN",
          "id": "19912156091999362_0",
          "equipment": "763",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "LIM",
          "airline": "LA",
          "fare_classes": "A",
          "cityFrom": "Barcelona",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": true,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-05-05T23:00:00.000Z",
          "utc_arrival": "2019-05-06T04:00:00.000Z",
          "local_departure": "2019-05-05T16:50:00.000Z",
          "utc_departure": "2019-05-05T14:50:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 1321,
          "operating_carrier": "LA",
          "cityTo": "Asunci처n",
          "vehicle_type": "aircraft",
          "flyFrom": "LIM",
          "id": "19912156091999362_1",
          "equipment": "320",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "ASU",
          "airline": "LA",
          "fare_classes": "N",
          "cityFrom": "Lima",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-06T05:04:00.000Z",
          "utc_arrival": "2019-06-06T09:04:00.000Z",
          "local_departure": "2019-06-06T00:29:00.000Z",
          "utc_departure": "2019-06-06T05:29:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 1317,
          "operating_carrier": "LA",
          "cityTo": "S찾o Paulo",
          "vehicle_type": "aircraft",
          "flyFrom": "ASU",
          "id": "19912156091999362_2",
          "equipment": "320",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "GRU",
          "airline": "LA",
          "fare_classes": "B",
          "cityFrom": "Asunci처n",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-06T13:40:00.000Z",
          "utc_arrival": "2019-06-06T16:40:00.000Z",
          "local_departure": "2019-06-06T10:30:00.000Z",
          "utc_departure": "2019-06-06T14:30:00.000Z"
        },
        {
          "refresh_timestamp": "2019-04-04T10:01:51.000Z",
          "bags_recheck_required": false,
          "return": 1,
          "flight_no": 8114,
          "operating_carrier": "LA",
          "cityTo": "Barcelona",
          "vehicle_type": "aircraft",
          "flyFrom": "GRU",
          "id": "19912156091999362_3",
          "equipment": "359",
          "combination_id": "19912156091999362",
          "fare_family": "",
          "flyTo": "BCN",
          "airline": "LA",
          "fare_classes": "A",
          "cityFrom": "S찾o Paulo",
          "last_seen": "2019-04-04T10:01:51.000Z",
          "guarantee": false,
          "fare_basis": "ALESL50R",
          "local_arrival": "2019-06-07T08:45:00.000Z",
          "utc_arrival": "2019-06-07T06:45:00.000Z",
          "local_departure": "2019-06-06T17:20:00.000Z",
          "utc_departure": "2019-06-06T20:20:00.000Z"
        },
        {
          "refresh_timestamp": "1970-01-01T00:00:00.000Z",
          "bags_recheck_required": true,
          "return": 1,
          "flight_no": 1739,
          "operating_carrier": null,
          "cityTo": "Oslo",
          "vehicle_type": "aircraft",
          "flyFrom": "BCN",
          "id": "4050602064459815_0",
          "equipment": null,
          "combination_id": "4050602064459815",
          "fare_family": "",
          "flyTo": "OSL",
          "airline": "DY",
          "fare_classes": "",
          "cityFrom": "Barcelona",
          "last_seen": "2019-04-04T06:10:05.000Z",
          "guarantee": true,
          "fare_basis": "",
          "local_arrival": "2019-06-07T14:35:00.000Z",
          "utc_arrival": "2019-06-07T12:35:00.000Z",
          "local_departure": "2019-06-07T11:10:00.000Z",
          "utc_departure": "2019-06-07T09:10:00.000Z"
        }
      ],
      "facilitated_booking_available": false,
      "countryFrom": {
        "code": "NO",
        "name": "Norway"
      },
      "local_arrival": "2019-05-05T23:00:00.000Z",
      "utc_arrival": "2019-05-06T04:00:00.000Z",
      "local_departure": "2019-05-05T07:00:00.000Z",
      "utc_departure": "2019-05-05T05:00:00.000Z"
    }
  ],
  "search_id": "d65d933b-541f-4880-b6a0-3f9ae4911c22"
}
