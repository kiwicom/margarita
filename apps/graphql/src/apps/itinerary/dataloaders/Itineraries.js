// @flow

import stringify from 'json-stable-stringify';
import qs from 'querystring';
import * as DateFNS from 'date-fns';
import { OptimisticDataloader } from '@kiwicom/graphql-utils';
import { UK_DATE_FORMAT } from '@kiwicom/margarita-config';

import fetch from '../../../services/fetch/tequilaFetch';
import {
  mapLocation,
  mapDate,
  getItineraryType,
  mapSectors,
} from '../helpers/Itineraries';
import type {
  ItinerariesReturnSearchParameters,
  ItinerariesOneWaySearchParameters,
  ApiResponseType,
  Itinerary,
} from '../Itinerary';

const stripTimeZoneOffset = (date: Date) =>
  DateFNS.addMinutes(date, date.getTimezoneOffset());

const parseDate = (date: Date) =>
  DateFNS.format(stripTimeZoneOffset(date), UK_DATE_FORMAT);

export const parseParameters = (
  input: ItinerariesReturnSearchParameters | ItinerariesOneWaySearchParameters,
) => {
  const { origin, destination, outboundDate } = input.itinerary;
  const inboundDate = input.itinerary.inboundDate
    ? input.itinerary.inboundDate
    : null;

  const flyFrom = origin.ids.join();
  const flyTo = destination && destination.ids ? destination.ids.join() : null;

  const commonSearchParams = {
    fly_from: flyFrom,
    ...(input.order && { asc: input.order === 'ASC' ? 1 : 0 }),
    ...(input.sort && { sort: input.sort }),
    date_from: parseDate(outboundDate.start),
    date_to: outboundDate.end ? parseDate(outboundDate.end) : null,
    fly_to: flyTo,
    ...(input.passengers && {
      adults: input.passengers.adults ?? 0,
      children: input.passengers.children ?? 0,
      infants: input.passengers.infants ?? 0,
    }),
    curr: 'EUR',
  };

  return {
    ...commonSearchParams,
    ...addReturnSearchQueryParams(inboundDate),
  };
};

const addReturnSearchQueryParams = inboundDate => {
  return {
    ...(inboundDate && {
      return_from: parseDate(inboundDate.start),
      ...(inboundDate.end && {
        return_to: parseDate(inboundDate.end),
      }),
    }),
  };
};

const fetchItineraries = async (
  parameters: $ReadOnlyArray<ItinerariesReturnSearchParameters>,
) => {
  const results: $ReadOnlyArray<ApiResponseType> = await Promise.all(
    parameters.map(params => {
      return fetch(`/v2/search?${qs.stringify(parseParameters(params))}`);
    }),
  );
  return results.map(res => {
    return sanitizeItineraries(res);
  });
};

const sanitizeItineraries = (response: ApiResponseType): Itinerary[] => {
  const itineraries = response.data;

  return itineraries.map(itinerary => {
    return {
      id: itinerary.id,
      type: getItineraryType(itinerary.routes),
      bookingToken: itinerary.booking_token,
      isValid: false,
      isChecked: false,
      startTime: mapDate(itinerary.local_departure, itinerary.utc_departure),
      endTime: mapDate(itinerary.local_arrival, itinerary.utc_arrival),
      destination: mapLocation(
        itinerary.flyTo,
        itinerary.cityTo,
        itinerary.countryTo?.name,
        itinerary.countryTo?.code,
      ),
      origin: mapLocation(
        itinerary.flyFrom,
        itinerary.cityFrom,
        itinerary.countryFrom?.name,
        itinerary.countryFrom?.code,
      ),
      sectors: mapSectors(itinerary.route, itinerary.routes),
      price: {
        currency: response.currency,
        amount: itinerary.price,
      },
    };
  });
};

export function createItinerariesLoader() {
  return new OptimisticDataloader(
    async (
      keys: $ReadOnlyArray<ItinerariesReturnSearchParameters>,
    ): Promise<Array<Itinerary[] | Error>> => fetchItineraries(keys),
    {
      cacheKeyFn: stringify,
    },
  );
}
