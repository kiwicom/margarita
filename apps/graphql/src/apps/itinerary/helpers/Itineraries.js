// @flow

import * as DateFNS from 'date-fns';
import { head, last } from 'ramda';
import { fromGlobalId } from '@kiwicom/graphql-global-id';

import type { ApiRouteItem, Sector, Segment } from '../Itinerary';
import type { RouteStop } from '../../booking/Booking';

export const differenceInMinutes = (
  from: ?(string | number),
  to: ?(string | number),
) => {
  if (from == null || to == null) {
    return null;
  }
  const parseNonSpecificDate = date =>
    typeof date === 'string'
      ? DateFNS.parseISO(date)
      : DateFNS.fromUnixTime(date);

  return parseInt(
    DateFNS.differenceInMinutes(
      parseNonSpecificDate(to),
      parseNonSpecificDate(from),
    ),
    10,
  );
};

export const mapToOpaqueID = (input: any) => {
  const parsedOriginIds: string[] = input.itinerary.origin.ids.map(id =>
    fromGlobalId(id),
  );
  const parsedDestinationIds: string[] = input.itinerary.destination.ids.map(
    id => fromGlobalId(id),
  );

  return {
    ...input,
    itinerary: {
      ...input.itinerary,
      origin: {
        ids: parsedOriginIds,
      },
      destination: {
        ids: parsedDestinationIds,
      },
    },
  };
};

export const mapVehicle = (type: ?string, uniqueNo: ?string) => ({
  type: type ?? null,
  uniqueNo: uniqueNo ?? null,
});

export const mapTransporter = (transporter: ?string) => ({
  name: transporter ?? null,
});

export const getItineraryType = (routes: ?Array<Array<string>>) => {
  if (routes == null) {
    return null;
  }
  if (routes.length === 1) {
    return 'oneway';
  }
  if (routes.length === 2) {
    return 'return';
  }
  return null;
};

/**
 * This function converts route and routes data from REST API to well-structured Sectors with Segments.
 * Sector = The whole part from source to destination, e.g Oslo -> Prague. A return trip will have 2 sectors
 * and 1 way trip will have 1 sector. A multicity trip could have 2 or more sectors.
 * Segment = A part of the sector. Segments could be Oslo -> Warzaw, Warzaw -> Prague
 */
export const mapSectors = (
  routesList: ?(ApiRouteItem[]),
  routeCodes: ?Array<string[]>,
): ?Array<Sector> => {
  if (routesList == null || routeCodes == null) {
    return null;
  }
  const sectors = [];
  let currentSectorIndex = 0;
  let currentRouteCode = head(routeCodes) ?? [];

  let { currentSector, currentArrival } = sanitizeSector(
    routesList,
    currentRouteCode,
  );

  sectors[currentSectorIndex] = currentSector;

  routesList.forEach((route: ApiRouteItem) => {
    const segment = sanitizeSegment(route);

    if (sectors[currentSectorIndex].segments) {
      sectors[currentSectorIndex].segments.push(segment);
    }

    if (route.flyTo === currentArrival?.flyTo) {
      currentSectorIndex++;
      // $FlowExpectedError: We already tested that routeCodes != null
      currentRouteCode = routeCodes[currentSectorIndex];

      if (currentRouteCode == null) {
        return;
      }
      // $FlowExpectedError: We already tested that routesList != null
      const nextSector = sanitizeSector(routesList, currentRouteCode);
      currentArrival = nextSector.currentArrival;
      sectors[currentSectorIndex] = nextSector.currentSector;
    }
  });

  return sectors;
};

const sanitizeSector = (
  routesList: ApiRouteItem[],
  currentRouteCode: string[],
) => {
  const currentRouteArrivalCode = last(currentRouteCode) ?? '';
  const currentRouteDepartureCode = head(currentRouteCode) ?? '';

  const currentDeparture: ?ApiRouteItem = routesList.find(
    route => route.flyFrom === currentRouteDepartureCode,
  );

  const currentArrival: ?ApiRouteItem = routesList.find(
    route => route.flyTo === currentRouteArrivalCode,
  );

  const departure = sanitizeRouteStop(
    apiRouteItemToDeparture(currentDeparture),
  );
  const arrival = sanitizeRouteStop(apiRouteItemToArrival(currentArrival));

  const currentSector = {
    departure,
    arrival,
    arrivalTime: {
      local: arrival.time?.local,
      utc: arrival.time?.utc,
    },
    departureTime: {
      local: departure.time?.local,
      utc: departure.time?.utc,
    },
    duration: differenceInMinutes(departure.time?.utc, arrival.time?.utc),
    destination: mapLocation(currentArrival?.flyTo, currentArrival?.cityTo),
    origin: mapLocation(currentDeparture?.flyFrom, currentDeparture?.cityFrom),
    segments: [],
    stopoverDuration: 0,
  };

  return {
    currentArrival,
    currentSector,
  };
};

const apiRouteItemToArrival = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityTo,
  code: routeItem?.flyTo,
  time: {
    utc: routeItem?.utc_arrival,
    local: routeItem?.local_arrival,
  },
});

const apiRouteItemToDeparture = (routeItem: ?ApiRouteItem) => ({
  cityName: routeItem?.cityFrom,
  code: routeItem?.flyFrom,
  time: {
    utc: routeItem?.utc_departure,
    local: routeItem?.local_departure,
  },
});

const sanitizeRouteStop = ({
  cityName,
  code,
  time,
}: {|
  +cityName: ?string,
  +code: ?string,
  +time: {| +utc: ?string, +local: ?string |},
|}): RouteStop => ({
  cityName,
  cityId: null,
  time,
  code,
});

const sanitizeSegment = (segment: ?ApiRouteItem): Segment => {
  return {
    duration: differenceInMinutes(segment?.utc_departure, segment?.utc_arrival),
    id: segment?.id,
    transporter: mapTransporter(segment?.airline),
    vehicle: mapVehicle(segment?.vehicle_type, String(segment?.flight_no)),
    departure: sanitizeRouteStop(apiRouteItemToDeparture(segment)),
    arrival: sanitizeRouteStop(apiRouteItemToArrival(segment)),
  };
};

export const mapDate = (local: ?string, utc: ?string) => {
  const localDate = local && DateFNS.parseISO(local).toISOString();
  const utcDate = utc && DateFNS.parseISO(utc).toISOString();
  return {
    local: localDate ?? null,
    utc: utcDate ?? null,
  };
};

export const mapLocationArea = (
  code: ?string,
  locationId: ?string,
  name: ?string,
  slug: ?string,
  flagURL: ?string,
) => ({
  id: locationId ?? null,
  code: code ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  slug: slug ?? null,
  flagURL: flagURL ?? null,
});

export const mapLocation = (
  locationId: ?string,
  name: ?string,
  countryName: ?string,
  countryCode: ?string,
  slug: ?string,
) => ({
  id: locationId ?? null,
  locationId: locationId ?? null,
  name: name ?? null,
  timezone: 'UTC+1',
  country: mapLocationArea(countryCode, countryCode, countryName),
  slug: slug ?? null,
  city: null,
  type: null,
});
