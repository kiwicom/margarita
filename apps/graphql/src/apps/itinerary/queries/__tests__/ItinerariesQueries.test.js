// @flow

import { generateTestsFromFixtures } from '@kiwicom/test-utils';

import executeTestQuery from '../../../../services/testingTools/executeTestQuery';
import Itineraries from '../../__datasets__/Itineraries.json';
import ItineraryReturn from '../../__datasets__/ItineraryReturn.json';

generateTestsFromFixtures(`${__dirname}/__fixtures__/itineraries`, input => {
  fetch.mockResponses(
    [JSON.stringify(Itineraries)],
    [JSON.stringify(ItineraryReturn)],
  );
  return executeTestQuery(input);
});
