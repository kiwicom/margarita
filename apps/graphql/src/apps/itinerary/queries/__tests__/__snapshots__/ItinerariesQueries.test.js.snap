// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches expected output: ItinerariesOneWay.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment RouteStopFragment on RouteStop {
  time {
    utc
    local
  }
}

{
  searchOneWayItineraries(input: {
  order: ASC,
  sort: QUALITY,
  passengers: {
    adults: 1,
    children: 0,
    infants: 0,
  },
  itinerary: {
    origin: {
      ids: ["prague_cz"],
    },
    destination: {
      ids: ["STN"],
    },
    outboundDate: {
      start: "2019-05-01",
      end: "2019-05-03",
    },
  },
}, first: 1) {
    edges {
      node {
        id
        sectors {
          duration
          stopoverDuration
          segments {
            departure {
              ...RouteStopFragment
            }
            arrival {
              ...RouteStopFragment
            }
          }
          departure {
            ...RouteStopFragment
          }
          arrival {
            ...RouteStopFragment
          }
        }
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "searchOneWayItineraries": {
      "edges": [
        {
          "node": {
            "id": "SXRpbmVyYXJ5OjM5OTEyMjc5NzgyMjAxMTlfMHwzOTE1MzYxNjAzMzExMTg1XzB8ODc3NjMwMTgzOTg2NDEyM18wfDEyMzQ1MzE2NjE2NTYxNjc4XzA=",
            "sectors": [
              {
                "duration": 3435,
                "stopoverDuration": null,
                "segments": [
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-01T17:45:00.000Z",
                        "local": "2019-03-01T18:45:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-01T18:50:00.000Z",
                        "local": "2019-03-01T19:50:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-02T14:15:00.000Z",
                        "local": "2019-03-02T15:15:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-03T00:45:00.000Z",
                        "local": "2019-03-02T19:45:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-03T03:55:00.000Z",
                        "local": "2019-03-02T22:55:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-03T07:24:00.000Z",
                        "local": "2019-03-03T02:24:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-03T23:50:00.000Z",
                        "local": "2019-03-03T18:50:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-04T03:00:00.000Z",
                        "local": "2019-03-03T22:00:00.000Z"
                      }
                    }
                  }
                ],
                "departure": {
                  "time": {
                    "utc": "2019-03-01T17:45:00.000Z",
                    "local": "2019-03-01T18:45:00.000Z"
                  }
                },
                "arrival": {
                  "time": {
                    "utc": "2019-03-04T03:00:00.000Z",
                    "local": "2019-03-03T22:00:00.000Z"
                  }
                }
              }
            ]
          }
        }
      ]
    }
  }
}
`;

exports[`matches expected output: ItinerariesReturn.graphql 1`] = `
~~~~~~~~~~ INPUT ~~~~~~~~~~
fragment RouteStopFragment on RouteStop {
  time {
    utc
    local
  }
}

{
  searchReturnItineraries(input: {
  order: ASC,
  sort: QUALITY,
  passengers: {
    adults: 1,
    children: 0,
    infants: 0,
  },
  itinerary: {
    origin: {
      ids: ["prague_cz"],
    },
    destination: {
      ids: ["STN"],
    },
    outboundDate: {
      start: "2019-05-01",
      end: "2019-05-03",
    },
    inboundDate: {
      start: "2019-05-25",
      end: "2019-05-28"
    }
  },
}, first: 1) {
    edges {
      node {
        id
        sectors {
          duration
          stopoverDuration
          segments {
            departure {
              ...RouteStopFragment
            }
            arrival {
              ...RouteStopFragment
            }
          }
          departure {
            ...RouteStopFragment
          }
          arrival {
            ...RouteStopFragment
          }
        }
      }
    }
  }
}
~~~~~~~~~~ OUTPUT ~~~~~~~~~~
{
  "data": {
    "searchReturnItineraries": {
      "edges": [
        {
          "node": {
            "id": "SXRpbmVyYXJ5OjM5OTEyMjc5NzgyMjAxMTlfMHwzOTE1MzYxNjAzMzExMTg1XzB8ODc3NjMwMTgzOTg2NDEyM18wfDEyMzQ1MzE2NjE2NTYxNjc4XzA=",
            "sectors": [
              {
                "duration": 3435,
                "stopoverDuration": null,
                "segments": [
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-01T17:45:00.000Z",
                        "local": "2019-03-01T18:45:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-01T18:50:00.000Z",
                        "local": "2019-03-01T19:50:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-02T14:15:00.000Z",
                        "local": "2019-03-02T15:15:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-03T00:45:00.000Z",
                        "local": "2019-03-02T19:45:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-03T03:55:00.000Z",
                        "local": "2019-03-02T22:55:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-03T07:24:00.000Z",
                        "local": "2019-03-03T02:24:00.000Z"
                      }
                    }
                  },
                  {
                    "departure": {
                      "time": {
                        "utc": "2019-03-03T23:50:00.000Z",
                        "local": "2019-03-03T18:50:00.000Z"
                      }
                    },
                    "arrival": {
                      "time": {
                        "utc": "2019-03-04T03:00:00.000Z",
                        "local": "2019-03-03T22:00:00.000Z"
                      }
                    }
                  }
                ],
                "departure": {
                  "time": {
                    "utc": "2019-03-01T17:45:00.000Z",
                    "local": "2019-03-01T18:45:00.000Z"
                  }
                },
                "arrival": {
                  "time": {
                    "utc": "2019-03-04T03:00:00.000Z",
                    "local": "2019-03-03T22:00:00.000Z"
                  }
                }
              }
            ]
          }
        }
      ]
    }
  }
}
`;
